File: software/examples/delphi/ExampleCallback.pas
program ExampleCallback;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, BrickletMultiTouch;

type
  TExample = class
  private
    ipcon: TIPConnection;
    mt: TBrickletMultiTouch;
  public
    procedure TouchStateCB(sender: TBrickletMultiTouch; const touchState: word);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;
  UID = 'XYZ'; { Change to your UID }

var
  e: TExample;

{ Callback function for touch state }
procedure TExample.TouchStateCB(sender: TBrickletMultiTouch; const touchState: word);
var i: integer;
begin
  if (touchState And (1 Shl 12)) = (1 Shl 12) then begin
    Write('In proximity, ');
  end;

  if (touchState And $fff) = 0 then begin
    WriteLn('No electrodes touched');
  end
  else begin
    Write('Electrodes ');
    for i:= 0 to 11 do
	begin
	  if (touchState And (1 Shl i)) = (1 Shl i) then begin
	    Write(IntToStr(i) + ' ');
      end;
	end;
	WriteLn('touched');
  end;
  WriteLn('');
end;

procedure TExample.Execute;
begin
  { Create IP connection }
  ipcon := TIPConnection.Create;

  { Create device object }
  mt := TBrickletMultiTouch.Create(UID, ipcon);

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);
  { Don't use device before ipcon is connected }

  { Register touch state callback to procedure TouchStateCB }
  mt.OnTouchState := {$ifdef FPC}@{$endif}TouchStateCB;

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.

File: software/examples/vbnet/ExampleCallback.vb
Imports System
Imports Tinkerforge

Module ExampleCallback
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223
    Const UID As String = "XYZ" ' Change XYZ to the UID of your Multi Touch Bricklet

    ' Callback subroutine for touch state callback
    Sub TouchStateCB(ByVal sender As BrickletMultiTouch, ByVal state As Integer)
        Dim s As String = ""

        If ((state And (1 << 12)) = (1 << 12)) Then
            s &= "In proximity, "
        End If

        If ((state And &Hfff) = 0) Then
            s &= "No electrodes touched"
        Else
            s &= "Electrodes "
            For i As Integer = 0 To 11
                If ((state And (1 << i)) = (1 << i)) Then
                    s &= i.ToString() & " "
                End If
            Next
            s &= "touched"
        End If

        Console.WriteLine(s)
    End Sub

    Sub Main()
        Dim ipcon As New IPConnection() ' Create IP connection
        Dim mt As New BrickletMultiTouch(UID, ipcon) ' Create device object

        ipcon.Connect(HOST, PORT) ' Connect to brickd
        ' Don't use device before ipcon is connected

        ' Register touch state callback to subroutine TouchStateCB
        AddHandler mt.TouchState, AddressOf TouchStateCB

        Console.WriteLine("Press key to exit")
        Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module
